src/main/java/Controller/UserAPI.java

```diff
@@ -0,0 +1,120 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.http.ResponseEntity;
+import javax.servlet.http.HttpServletRequest; 
+
+import java.util.List;
+import java.util.Map;
+
+@RestController
+public class UserAPI {
+
+    @Autowired 
+    private UserAccessor userAccessor;
+
+    @Autowired
+    private UniqueIDGenerator uniqueIDGenerator;
+
+    @Autowired
+    private MissingFields missingFields;
+
+    @Autowired
+    private ErrorResponseBody errorResponseBody;
+
+    @ExceptionHandler(UserAlreadyExistsException.class)
+    public ResponseEntity<Object> handleUserAlreadyExistsException(UserAlreadyExistsException e) {
+        return new ResponseEntity<Object>(errorResponseBody.getResponseBody(e), e.getHttpStatus());
+    }  
+     
+    @ExceptionHandler(UserRequiredFieldMissingException.class)
+    public ResponseEntity<Object> handleRequiredFieldMissingException(UserRequiredFieldMissingException e) {
+        return new ResponseEntity<Object>(errorResponseBody.getResponseBody(e), e.getHttpStatus());
+    }
+
+    @ExceptionHandler(InvalidLoginException.class)
+    public ResponseEntity<Object> handleInvalidLoginException(InvalidLoginException e) {
+        return new ResponseEntity<Object>(errorResponseBody.getResponseBody(e), e.getHttpStatus());
+    }
+
+    @ExceptionHandler(UserNotFoundException.class)
+    public ResponseEntity<Object> handleUserNotFoundException(UserNotFoundException e) {
+        return new ResponseEntity<Object>(errorResponseBody.getResponseBody(e), e.getHttpStatus());
+    }
+
+    @PostMapping(value = "/signup", consumes={"application/json"})
+    public Map<String, Object> signup(@RequestBody Map<String, Object> data, HttpServletRequest request) {
+
+        String path = request.getRequestURI();
+        String[] required = {"Username", "EmailID", "Password", "MobileNo"};
+        List<String> missing_fields = missingFields.getMissingFields(data, required);
+        if( missing_fields.size() > 0 ) {
+            throw new UserRequiredFieldMissingException(path, missing_fields);
+        }
+
+        /* necessary fields */
+        String username = data.get("Username").toString();
+        String emailID = data.get("EmailID").toString();
+        String password = data.get("Password").toString();
+        String mobileNo = data.get("MobileNo").toString();
+
+        /* optional field */
+        String base64Image = "";
+        if(data.get("Picture") != null){
+            base64Image = data.get("Picture").toString();
+        }
+
+        /* Checks if Username or email exists - throws exception if it does */
+        if (userAccessor.checkIfUserExists(username, emailID)) {
+            throw new UserAlreadyExistsException(path);
+        }
```

Comment:
Would it be better to return the user object, rather than throwing exception here (as we discussed in the last meeting) ?
@sankalpbansal-google , @AgamAgarwal WDYT ?

Reply:
It might happen that a new user gives in a username that is already taken by some other user, so in that case we should not return the user object. However, if `username`, `emailID`, `mobileNo`, `Password` all happen to match then I think we can return the user object.

Reply:
What do you suggest I should do?